test/zimfile.cpp

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2020 Veloman Yunkan
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * is provided AS IS, WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, and
+ * NON-INFRINGEMENT.  See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ */
+
+#include <zim/zim.h>
+#include <zim/file.h>
+
+#include "tempfile.h"
+#include "../src/fs.h"
+
+#include "gtest/gtest.h"
+
+namespace
+{
+
+using zim::unittests::TempFile;
+
+using TestContextImpl = std::vector<std::pair<std::string, std::string> >;
+struct TestContext : TestContextImpl {
+  TestContext(const std::initializer_list<value_type>& il)
+    : TestContextImpl(il)
+  {}
+};
+
+std::ostream& operator<<(std::ostream& out, const TestContext& ctx)
+{
+  out << "Test context:\n";
+  for ( const auto& kv : ctx )
+    out << "\t" << kv.first << ": " << kv.second << "\n";
+  out << std::endl;
+  return out;
+}
+
+std::string
+emptyZimFileContent()
+{
+  std::string content;
+  content += "ZIM\x04"; // Magic
+  content += "\x05" + std::string(3, '\0'); // Version
+  content += std::string(16, '\0'); // uuid
+  content += std::string(4, '\0'); // article count
+  content += std::string(4, '\0'); // cluster count
+  content += "\x50" + std::string(7, '\0'); // url ptr pos
+  content += "\x50" + std::string(7, '\0'); // title ptr pos
+  content += "\x50" + std::string(7, '\0'); // cluster ptr pos
+  content += "\x50" + std::string(7, '\0'); // mimelist ptr pos
+  content += std::string(4, '\0'); // main page index
+  content += std::string(4, '\0'); // layout page index
+  content += "\x50" + std::string(7, '\0'); // checksum pos
+  content += "\x8a\xbb\xad\x98\x64\xd5\x48\xb2\xb9\x71\xab\x30\xed\x29\xa4\x01"; // md5sum
+  return content;
+}
+
+class TempZimFile : public TempFile
+{
+public:
+  TempZimFile(const char* name, const std::string& content)
+    : TempFile(name)
+  {
+    write(fd(), &content[0], content.size());
+    close();
```

Comment:
On windows, closing the fd also delete the file. So it leads to an unusable `TempZimFile`.

Reply:
Good catch!

There are several ways to fix that. In my opinion, the easiest at this point is to undo the commit bc844b7 (which introduced `TempFile::close()`), while preserving the API restriction that the file descriptor returned by `TempFile::fd()` **must not** be `close()`-ed. Is it OK?

Reply:
Yes, seems good to remove the `close` method. It should not be a problem to keep the file opened a bit longer.

Reply:
Done. I added an extra commit 4ce4e02 that properly documents `TempFile`.